### ISSUE TITLE
RFC: Multidimensional (cubature) integration

### ISSUE THREAD

**Issue by izaid** [scipy/scipy#20252]
Hello all! This issue is seeking thoughts on a core part of `scipy.integrate`. More than that, it is proposing an enhancement that I and others are willing and able to work towards.

Integrating a function $f(x)$ is a fundamental part of numerical computing, and doing so in $n$ dimensions is just as important as doing so in 1 dimension. SciPy currently has routines in its `integrate` submodule to do so, notably `dblquad`, `tplquad`, and `nquad`. These are wrappers that essentially do 1D iterated integrals using `quad` from the Fortran library QUADPACK. QUADPACK is very impressive software, but at the end of the day it is still written in Fortran and limited by the design of its time. The version in SciPy is from 1984.

Improving `integrate` generally is important, and I'd very much like feedback on a plan on how to do that for quadrature in $n$ dimensions. If what is written here is broadly acceptable to everyone, I and others would intend to carry out this work on multidimensional integration before the end of this summer. The inspiration here is really the adaptive quadrature in `quad_vec`, which I think went a long way to providing robust integration in SciPy that works well with arrays.

To be very specific about the issues, multidimensional integration in SciPy currently has these limitations:

- It does not support array-valued integrands of arbitrary shape, only scalar functions. Therefore one cannot integrate a vector or matrix or tensor function whereas `quad_vec` can.

- It does not support vectorization. If one wants to speed things up by sampling all the integration points at once, you can't.

- It will never support other array APIs, as it is written in Fortran. So it can never go on a GPU.

- It does not support custom integration rules. If one wants to use, say, multidimensional tanh-sinh quadrature for handling a singular point, you can't.

The proposal here is to provide a multidimensional integration function that is adaptive and similar to `quad_vec`. This is what is known as "cubature", we can call it `cub` or `ndquad_vec` or whatever. It will take almost the same arguments as `quad_vec`, except `a`, `b`, and `points` will all become $n$ dimensional. It will use the array API to enable support for GPU integration and beyond.

An additional feature that it will have, which `quad_vec`doesn't have, is support for passing in a custom integration rule. This would allow the adaptive part to subdivide effectively, but it could then use the custom integration rule to calculate the integral in each subdivision. The built-in rules would be Gauss-Kronrod and Genz-Malik (popular for multidimensional integrands). But users could also pass in rules that effectively deal with singular (using tanh-sinh) or oscillatory integrands. QUADPACK actually has behaviour like this with its methods "qawoe", "qawse", and "qawce", but it is not fully custom. If successful, this behaviour could be added into `quad_vec` as well by simply making its existing `quadrature` argument also accept a callable.

This proposal would go a long way towards improving an existing capability within SciPy. It would also help a lot with removing Fortran from the codebase.

I know @mdhaber, among others, has been doing work on integrate. I'd very much appreciate his thoughts and, of course, everyone's.

**Comment by andyfaff** [comment]
I would think inclusion of LowLevelCallable would be important.

**Comment by ev-br** [comment]
A strong +1 from me in general. 

Details:

- is the proposal to only provide a new `ndim > 1` routine, or is bringing `quad_vec` to feature completeness with `quad` a part of it? 
- the interface needs careful design, esp around vectorization, vector-valued functions and broadcasting w.r.t. parameters. IIUC, the current `_tanhsinh` is incompatible with `quad_vec`. This is somewhat resembling the situation we have with `approx_derivative` and `misc.derivative`: vector-matrix semantics is at odds with broadcasting. It would be best to avoid the split, if at all possible. Or else, provide multiple interfaces with shared underlying integration machinery.
- custom integration rules are great. Now, would be great to have a stance on a Monte Carlo rule: are you open to a possibility of some sort of an MC integration rule or does it constrain the interface or internals? Not that I've an opinion; it's just bound to show up once somebody want ndim > 5 or so, and it's best to have some plan. Also once MC is there, QMC comes right after it. 
- A single `nquad`-like interface for general n-dimensional functions, or a collection of specializations for 2D, 3D etc?

**Comment by izaid** [comment]
> A strong +1 from me in general.

Thanks @ev-br! Some answers.

> Details:
> 
>     * is the proposal to only provide a new `ndim > 1` routine, or is bringing `quad_vec` to feature completeness with `quad` a part of it?

The proposal is to provide a new routine for `ndim >= 1`. As you point out, there are design considerations there to make sure we get it right. If we do, a second part would be to adapt `quad_vec` so it is fully compatible. I think `quad_vec`, or any successfor thereof, should remain a one-dimensional routine. For the sake of generality, I think `ndquad_vec` should also support one dimension, but it's intention is to be used for arbitrary $n$.

>     * the interface needs careful design, esp around vectorization, vector-valued functions and broadcasting w.r.t. parameters. IIUC, the current `_tanhsinh` is incompatible with `quad_vec`. This is somewhat resembling the situation we have with `approx_derivative` and `misc.derivative`: vector-matrix semantics is at odds with broadcasting. It would be best to avoid the split, if at all possible. Or else, provide multiple interfaces with shared underlying integration machinery.

Couldn't agree more! What I have described has already been prototyped and I see clearly all the issues you have raised.

>     * custom integration rules are great. Now, would be great to have a stance on a Monte Carlo rule: are you open to a possibility of some sort of an MC integration rule or does it constrain the interface or internals? Not that I've an opinion; it's just bound to show up once somebody want ndim > 5 or so, and it's best to have some plan. Also once MC is there, QMC comes right after it.

I don't see any reason why MC integration rules should be excluded. And indeed, we should talk to make sure we are as compatible for MC as possible! But I don't think we specifically would write any MC rules as part of this, that would be for someone else or for the future.

The design for each rule should be general. I'm essentially thinking of them as callables with the signature `__call__(self, func, a, b)`. If they need to calculate points and weights or whatever, they can do that and store them internally.

Really, the purpose here is to make the numerically integration *composable*.

>     * A single `nquad`-like interface for general n-dimensional functions, or a collection of specializations for 2D, 3D etc?

A single `nquad`-like interface for general n-dimensional functions. Drop the 2D and 3D ones.

**Comment by mdhaber** [comment]
> IIUC, the current _tanhsinh is incompatible with quad_vec

No, it now supports vector valued integrands. (The interface is not identical, though, if that's what is meant.)

> Also once MC is there, QMC comes right after it.

There is already [`qmc_quad`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.integrate.qmc_quad.html).

Perhaps the point was to have a consistent interface for many methods. Sure - the only reason `_tanhsinh` is not public yet is that I didn't want to introduce yet another functions (e.g. `tanhsinh`) to fragment `integrate`. The hope of deprecating `quadrature` and `romberg` was to unify things.

**Comment by mdhaber** [comment]
I read through this more carefully, and it looks great! Since you mention that the inspiration is `quad_vec`, I thought I'd check - have you looked at [`_tanhsinh`](https://github.com/scipy/scipy/blob/main/scipy/integrate/_tanhsinh.py)? Based on benchmarks in the original PR (gh-18650), `_tanhsinh` seems to be faster and more accurate than `quad_vec`, and for integration in one dimension (because the immediate goal was to improve the univariate distribution infrastructure), `_tanhsinh` was written specifically to address most of what is suggested here.

- It supports vectorization in the sense of sampling all the integration points at once.
- It *also* supports vectorization in the sense of allowing the limits of integration to be of arbitrary shape. (Of course, there are faster ways of computing indefinite integrals of univariate functions - and we have plans to implement those later - but it's a handy feature for now.)
- It now supports array-valued integrands of any shape, which was `quad_vec`'s claim to fame. (Enabling this support was pretty straightforward (gh-19545) because the fundamental restriction that prevented array-valued functions was *added* for efficiency with SISO functions; I just had to add an option to remove that restriction.)
- Although it is not yet Array-API compatible, that has always been the goal, and @tupui and I have specific plans for implementation in the coming months.
- Only one integration rule is implemented right now, but I think it will be straightforward to add others, and we discussed exposing several rules in a common interface in gh-19510. Note that this quadrature routine itself leverages a more general [elementwise iterative methods framework](https://github.com/scipy/scipy/blob/main/scipy/_lib/_elementwise_iterative_method.py). The framework takes care of a lot of the "bookkeeping" of writing such functions, and it has also been used to write differentiation, root-finding, and minimization routines. It would be great if you would be interested in improving/extending this framework to support functions with N-dimensional input.

Still, I didn't want to expose `_tanhsinh` publicly via the `quad_vec` function because that interface (e.g. accepting a `full_output` parameter, returning a `tuple`) is not the direction SciPy has been headed in. Also, `quad_vec` doesn't support several of `_tanhsinh`'s features (e.g. computing the `log` of the integral from the log of the integrand), and some of `quad_vec`'s options would not make sense for `_tanhsinh`, so it would get a bit confusing to try to combine them. We need a new interface designed with more flexibility in mind. I'm won't be teaching this spring or summer, so I had planned to work on such an interface this spring. It sounds like our timelines align well!

Did you intend to get into specific interface discussions here?

**Comment by zachjweiner** [comment]
There's a distinction between nested/composed 1D quadrature and "cubature," algorithms for which treat all dimensions of the domain at once. The standard implementation of non-MC algorithms (to my knowledge) is [`cubature`](https://github.com/stevengj/cubature), which is GPL'ed; but the author has an MIT'ed [Julia port](https://github.com/JuliaMath/HCubature.jl) of the more commonly used algorithm, which adaptively subdivides the full domain along whatever dimension contributes the largest error at each iteration. Such methods are really necessary to efficiently integrate in dimensions higher than 2 or 3.

Some of the discussion suggests to me the proposal is restricted to nested 1D quadrature - is this the case, or is the proposal actually to implement algorithms for n-dimensional adaptivity?

I do think that true cubature algorithms would be a highly valuable addition to SciPy (and fill a bit of a whole in the ecosystem) and are established/standard enough to meet SciPy's bar for inclusion. Its interface, however, would surely differ from `nquad`-alikes (in its optional arguments, at least). I would recommend reserving method names deriving from "cubature" for them.

(To be clear, nested quadrature can be more efficient than, e.g., h-adaptive cubature - my memory is that it's often the case in 2D and maybe 3D, too. So I wouldn't drop `nquad` or an equivalent replacement in any case.)

**Comment by mdhaber** [comment]
My understanding is that @izaid is proposing "true cubature". I proposed nested quadrature with `_tanhsinh` before, but yes, cubature would be better for N-D.

**Comment by izaid** [comment]
Thanks @mdhaber and @zachjweiner for the further comments. My thoughts are as follows.

@mdhaber, I am aware of `tanh-sinh` and that's exactly something badly needed (quadratures for handling singular points). I think it's great! Here I am thinking of a general adaptive scheme, with plug-and-play quadratures, of which `tanh-sinh` would be one of.

An important point you mentioned, which I forgot to, is:

> It also supports vectorization in the sense of allowing the limits of integration to be of arbitrary shape. 

Indeed, and I want that supported here as well. That's important for doing families of integrals. I'm certainly happy to check out the elementwise iterative framework, and that may be very useful! In terms of interface, basically we should discuss in more detail. My experiments so far have led me to think that just having quadratures of the type `some_quad(func, a, b, **kwargs)` is potentially enough, where `some_quad` could also be a callable class if data is needed to be stored. But we should discuss.

@zachjweiner Yes, I'm discussing true cubature here. But cubature that may be composed, so you have an adaptive cubature scheme but the inner rule is Genz-Malik or tanh-sinh or whatever. BTW Genz-Malik is a true cubature scheme, in that it's not a product of 1D quadratures.

Think that's all the questions for now?

**Comment by lorentzenchr** [comment]
FYI, a few years ago the Cuba library in C was one of the best multidim quadrature libs. It has one (deterministic) curbature rule called Cuhre that I‘ve used a lot in the past. (But is LGPL licensed.)

**Comment by izaid** [comment]
Thanks @lorentzenchr! Yes, I am aware of Cuba and the deterministic Cuhre rule, but good to keep it in mind.

@ollybritton has been prototyping this proposal, along with myself and @ev-br. We hope to share a PR soon.

**Comment by lucascolley** [comment]
@ollybritton @izaid how does this issue stand after the merge of gh-21330? Perhaps you could describe what work remains before this should be closed?

**Comment by izaid** [comment]
> @ollybritton @izaid how does this issue stand after the merge of [gh-21330](https://github.com/scipy/scipy/pull/21330)? Perhaps you could describe what work remains before this should be closed?

@ollybritton, what do you think?

**Comment by lucascolley** [comment]
The cubature docs attempt to link to `CubatureResult`, but there is no link since that isn't part of the public API.

<img width="708" alt="image" src="https://github.com/user-attachments/assets/43d025a8-898c-410c-aed0-8ca15198bb5f">

If it is going to stay private, should we just explain the fields of `CubatureResult` here under `Returns`?

**Comment by ollybritton** [comment]
@lucascolley @izaid The original 4 problems that wanted to be addressed were:

> - It does not support array-valued integrands of arbitrary shape, only scalar functions. Therefore one cannot integrate a vector or matrix or tensor function whereas quad_vec can.
> - It does not support vectorization. If one wants to speed things up by sampling all the integration points at once, you can't.
> - It will never support other array APIs, as it is written in Fortran. So it can never go on a GPU.
> - It does not support custom integration rules. If one wants to use, say, multidimensional tanh-sinh quadrature for handling a singular point, you can't.

and I think #21330 and #21473 cover all of these points. For custom integration rules, the API for this is currently private but could be made public in the future.

Another thing mentioned was support for `LowLevelCallable` which  `cubature` doesn't currently have, so this is something that could be implemented in the future. Not discussed above is integration over non-rectangular regions using functions as limits like `nquad`. Neither `cubature` or `quad_vec` support this but I'll make a PR for adding this to `cubature` soon.

I'll also make a pull request fixing the `CubatureResult` docs issue. If anything mentioned above has slipped through the cracks then please let me know!

**Comment by lucascolley** [comment]
I think the other thing that was mentioned above was including an interface to `_tanhsinh` in `cubature`. Is that something that would be possible to add in the future?

**Comment by mdhaber** [comment]
I think it was mentioned that the tanh-sinh rule could be implemented, not that `_tanhsinh` itself would be exposed. `_tanhsinh` itself is 1-D and has a different set of features.

**Comment by lucascolley** [comment]
Right. Are you any closer to figuring out how we should expose `_tanhsinh` then?

**Comment by izaid** [comment]
> I think it was mentioned that the tanh-sinh quadrature rule could be implemented, not that `_tanhsinh` itself would be exposed. `_tanhsinh` itself is 1-D and has a different set of features.

Yep! tanh-sinh can be implemented as a particular rule. By doing so, it can be used adaptively. I think this is good thing to do, but was not something we had time to get around to this summer.

I think there is now also a longer term question of whether updates should be done to `quad_vec` to bring it in line with `cubature`, effectively becoming a new `quadrature`. It should be equivalent to what `cubature` is currently doing in 1D.

**Comment by ollybritton** [comment]
> @lucascolley @izaid The original 4 problems that wanted to be addressed were:
> 
> > * It does not support array-valued integrands of arbitrary shape, only scalar functions. Therefore one cannot integrate a vector or matrix or tensor function whereas quad_vec can.
> > * It does not support vectorization. If one wants to speed things up by sampling all the integration points at once, you can't.
> > * It will never support other array APIs, as it is written in Fortran. So it can never go on a GPU.
> > * It does not support custom integration rules. If one wants to use, say, multidimensional tanh-sinh quadrature for handling a singular point, you can't.
> 
> and I think #21330 and #21473 cover all of these points. For custom integration rules, the API for this is currently private but could be made public in the future.
> 
> Another thing mentioned was support for `LowLevelCallable` which `cubature` doesn't currently have, so this is something that could be implemented in the future. Not discussed above is integration over non-rectangular regions using functions as limits like `nquad`. Neither `cubature` or `quad_vec` support this but I'll make a PR for adding this to `cubature` soon.
> 
> I'll also make a pull request fixing the `CubatureResult` docs issue. If anything mentioned above has slipped through the cracks then please let me know!

The pull request for non-rectangular regions of integration (#21632) has now been split into two, and the first of these (#21657) has been merged but I haven't yet opened the PR for supporting functions as limits. Still remaining from this batch of work:

- PR for supporting functions as limits.
- Add a section about `cubature` to the release notes.
- Update user guide for `integrate` package to include a guide to `cubature`.
- Add a test for when an unknown rule is specified as an argument to `cubature`: <https://github.com/scipy/scipy/pull/21657#discussion_r1790525526>.

I'll address these over the next few days. As well as everything mentioned above, some further follow-ups (potentially detailed in a new issue): 

- Support for infinite limits and functions as limits means that JAX is not supported because the implementation uses indexed assignment,
- Calculate Gauss-Kronrod nodes rather than hard-coding them,
- Expose the currently private `rules` API to allow for custom plug-and-play rules in `cubature`,
- Add more specialised rules to the rules API to support a wider class of integrals,
- Add support for the rules API to other functions in `integrate`, for example by allowing the `method` argument of `quad_vec` to accept a `Rule`,
- Improve performance by using more accurate error estimates.

**Comment by ev-br** [comment]
One other relatively small item is that `cubature` tests seem to extremely slow under pytorch. Might be worth looking into it.

**Comment by alphataubio** [comment]
@izaid @mdhaber @tupui im debugging scipy.integrate.cubature with multiprocessing (workers=-1). grok spotted the issue deep in scipy.integrate internals:

> The error TypeError: cannot pickle 'module' object points to an attempt to serialize a module, such as numpy. Examining the GaussKronrodQuadrature class (from SciPy’s _rules module), we find that it stores the array module as an attribute, self.xp = xp, where xp is typically numpy. Python modules are not picklable, so when the multiprocessing Pool tries to serialize the rule object, it fails due to this xp attribute.

this patch resolved the bug but it seems like a temporary hack (works for me). why is scipy architectured so that a different module than numpy might be passed in ? 

```
# Patch for scipy/integrate/_rules/_base.py to make Rule objects picklable
# Add these methods to the Rule class

def __getstate__(self):
    state = self.__dict__.copy()
    state.pop('xp', None)  # Exclude the xp attribute (a module) during pickling
    return state

def __setstate__(self, state):
    self.__dict__.update(state)
    self.xp = __import__('numpy')  # Reconstruct xp as NumPy on unpickling
```

**Comment by lucascolley** [comment]
@alphataubio please open a new issue rather than re-using this one. Happy to discuss my responses further over in a new issue.

---

>  cannot pickle 'module' object points to an attempt to serialize a module

This came up recently elsewhere and there are possible workarounds, e.g. https://github.com/scipy/scipy/blob/56e3c8d6153cde6e40cfcc6ca95cce318d800ebb/scipy/spatial/transform/_rigid_transform.py#L1698-L1708 was introduced recently over in `scipy.spatial`.

> why is scipy architectured so that a different module than numpy might be passed in ?

please see https://github.com/scipy/scipy/issues/18286
